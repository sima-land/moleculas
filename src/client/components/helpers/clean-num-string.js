import pipe from 'lodash/fp/pipe';
import toString from 'lodash/toString';

/**
 * Очищает строку от любых символов, кроме цифр и удаляет лишние нули вначале.
 * @param {string} str Строка.
 * @return {string} Возвращает очищенную строку.
 */
export const cleanNumString = pipe(
  str => toString(str) && str.replace(/[^0-9]/g, ''),
  str => str === '' ? '' : Number(str).toString()
);
