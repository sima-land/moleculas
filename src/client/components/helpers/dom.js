import isNumber from 'lodash/isNumber';

/**
 * Обрезает textContent переданного элемента, пока его высота не будет соответствовать заданной.
 * @param {HTMLElement} element Элемент для обрезания textContent.
 * @param {number} height Заданная высота.
 */
export const cutTextContent = (element, height) => {
  if (element instanceof HTMLElement && isNumber(height) && height > 0) {
    const selfDisplay = element.style.display;

    element.style.display = 'block';
    while (element.textContent.length > 3 && element.clientHeight > height) {
      // удаляем 4 элемента и вставляем многоточие
      element.textContent = `${element.textContent.slice(0, -4).trim()}...`;
    }
    element.style.display = selfDisplay;
  }
};
