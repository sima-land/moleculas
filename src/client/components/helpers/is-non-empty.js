/**
 * Определяет, что полученный аргумент - не пустой массив.
 * @param {Array} data Данные.
 * @return {boolean} Если массив с данными - true и false если не массив или пустой массив.
 */
export const isNonEmptyArray = data => Array.isArray(data) && data.length > 0;
