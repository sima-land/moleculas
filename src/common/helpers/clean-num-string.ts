/**
 * Очищает строку от любых символов, кроме цифр и удаляет лишние нули вначале.
 * @param value Строка.
 * @return Возвращает очищенную строку.
 */
export const cleanNumString = (value: unknown): string => {
  const string = String(value).replace(/\D/g, '');
  return string.length > 0 ? Number(string).toString() : '';
};
