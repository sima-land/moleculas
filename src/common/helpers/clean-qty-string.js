/**
 * Отчищает строку от лишних символов и обрезает от лишних.
 * Если количество не является целым - оставляет только десятичную долю.
 * @param {string} qty Количество товара.
 * @param {boolean} canBeFloat Признак - количество товара может быть float значением.
 * @return {string} Количество товара.
 */
export const cleanQtyString = (qty = '', canBeFloat = false) => {
  const qtyWithDot = String(qty).replace(',', '.');
  let nextQty = '';
  if (!Number.isNaN(parseFloat(qtyWithDot))) {
    const dotIndex = qtyWithDot && qtyWithDot.indexOf('.');
    const isFloatQty = canBeFloat && dotIndex !== -1;
    const qtyLength = isFloatQty ? 6 : 5;
    nextQty = String(parseFloat(qtyWithDot)).slice(0, qtyLength);

    const isFloatWithoutDot = isFloatQty && nextQty.indexOf('.') === -1;
    nextQty = isFloatWithoutDot ? `${nextQty}.` : nextQty;

    const hasDotInEndOfFullStr = nextQty.indexOf('.') === nextQty.length - 1 && nextQty.length === 6;
    nextQty = hasDotInEndOfFullStr
      ? nextQty.slice(0, nextQty.length - 1)
      : nextQty;

    // если float режем до десятичной дроби.
    nextQty = isFloatQty ? nextQty.slice(0, dotIndex + 2) : nextQty;
  }
  return nextQty.replace('.', ',');
};
