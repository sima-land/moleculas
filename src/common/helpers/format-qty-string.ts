/**
 * Отчищает строку от лишних символов и обрезает от лишних.
 * Если количество не является целым - оставляет только десятичную долю.
 * @param qty Количество товара.
 * @param options Опции.
 * @param options.allowFloat Признак - количество товара может быть float значением.
 * @return Количество товара.
 */
export const formatQtyString = (qty = '', { allowFloat = false }: { allowFloat?: boolean } = {}) => {
  const qtyWithDot = String(qty).replace(',', '.');
  let nextQty = '';

  if (!Number.isNaN(parseFloat(qtyWithDot))) {
    const dotIndex: number = qtyWithDot.indexOf('.');
    const isFloatQty: boolean = allowFloat && dotIndex !== -1;
    const qtyLength: number = isFloatQty ? 6 : 5;

    nextQty = String(parseFloat(qtyWithDot)).slice(0, qtyLength);

    const isFloatWithoutDot = isFloatQty && nextQty.indexOf('.') === -1;
    nextQty = isFloatWithoutDot ? `${nextQty}.` : nextQty;

    const hasDotInEndOfFullStr = nextQty.indexOf('.') === nextQty.length - 1 && nextQty.length === 6;
    nextQty = hasDotInEndOfFullStr
      ? nextQty.slice(0, nextQty.length - 1)
      : nextQty;

    // если float режем до десятичной дроби.
    nextQty = isFloatQty ? nextQty.slice(0, dotIndex + 2) : nextQty;
  }

  return nextQty.replace('.', ',');
};
