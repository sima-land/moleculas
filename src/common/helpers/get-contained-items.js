import isElement from 'lodash/isElement';

/**
 * Возвращает количество HTML элементов, размещенных по горизонтали в родительском элементе.
 * @param {HTMLElement} node DOM Элемент.
 * @return {number} Количество дочерних элементов.
 */
const getContainedItems = node => {
  let qty = 0;
  if (isElement(node)) {
    const childs = node.childNodes && Array.from(node.childNodes);
    const farItemIndex = childs.findIndex((item, index, arr) =>
      !arr[index + 1] || item.getBoundingClientRect().right > arr[index + 1].getBoundingClientRect().right);
    qty = farItemIndex + 1;
  }
  return qty;
};

export default getContainedItems;
