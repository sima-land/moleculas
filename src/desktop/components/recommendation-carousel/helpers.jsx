import { useEffect, useRef, useState } from 'react';
import on from '@dev-dep/ui-nucleons/helpers/on';

/**
 * Управляет расположением и появлением тултипа.
 * @return {Object} Объект с рефом и функциями для управления тултипом.
 */
export const useControlTooltip = () => {
  const [canShowTooltip, setCanShowTooltip] = useState(true);
  const tooltipRef = useRef();

  /**
   * Скрывает тултип.
   */
  const hideTooltip = () => {
    if (tooltipRef.current) {
      tooltipRef.current.style.opacity = 0;
    }
  };

  useEffect(() => on(document, 'mousedown', () => {
    setCanShowTooltip(false);
    hideTooltip();
  }), []);

  useEffect(() => on(document, 'mouseup', () => setCanShowTooltip(true)), []);

  /**
   * Происходит при наведении на кнопку быстрого просмотра.
   * @param {Event} event Событие при наведении на кнопку быстрого просмотра.
   */
  const onHoverQuickViewButton = event => {
    if (canShowTooltip && tooltipRef.current) {
      const { left, top } = event.currentTarget.getBoundingClientRect();
      tooltipRef.current.style.right = `calc(100% - ${left - 8}px)`; // 8px - отступ от иконки.
      tooltipRef.current.style.top = `${top + window.scrollY + 12}px`; // 12px - половина никонки.
      tooltipRef.current.style.opacity = 1;
    }
  };

  return {
    tooltipRef,
    onHoverQuickViewButton,
    onLeaveQuickViewButton: hideTooltip,
  };
};
