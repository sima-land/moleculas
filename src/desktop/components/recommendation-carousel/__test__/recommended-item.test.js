import React from 'react';
import { mount, shallow } from 'enzyme';
import RecommendedItem from '../recommended-item';
import Link from '@dev-dep/ui-nucleons/link';
import Price from '@dev-dep/ui-nucleons/price';
import { cutTextContent } from '../../../../common/helpers/dom';

import debounce from 'lodash/debounce';

jest.mock('lodash/debounce', () => jest.fn(fn => fn));

jest.mock('../../../../common/helpers/dom', () => {
  const original = jest.requireActual('../../../../common/helpers/dom');

  return {
    ...original,
    __esModule: true,
    cutTextContent: jest.fn(original.cutTextContent),
  };
});

describe('<RecommendedItem />', () => {
  describe('render with props', () => {
    const onItemFn = jest.fn();
    const onQuickViewFn = jest.fn();
    const onHoverQuickViewButtonFn = jest.fn();
    const onLeaveQuickViewButtonFn = jest.fn();
    const addGlobalListener = jest.fn();
    const props = {
      name: 'Милая картина из шерсти Schrodinger cat, А5',
      url: 'https://www.sima-land.ru/3266880/nabor-dlya-tvorchestva-topiariy-xyezhnost-d-shara-6-cm/',
      imageSrc: 'https://cdn2.static1-sima-land.com/items/3592229/4/140.jpg?v=1549604912',
      price: 135.4,
      currencyGrapheme: '$',
      onItemClick: onItemFn,
      onHoverQuickViewButton: onHoverQuickViewButtonFn,
      onLeaveQuickViewButton: onLeaveQuickViewButtonFn,
      addGlobalListener,
    };
    const allProps = {
      ...props,
      oldPrice: 234.9,
      onQuickViewClick: onQuickViewFn,
    };

    beforeEach(() => {
      onItemFn.mockClear();
      onQuickViewFn.mockClear();
      addGlobalListener.mockClear();
    });

    it('onItemClick and onQuickViewClick calls correctly', () => {
      jest.useFakeTimers();

      const component = mount(<RecommendedItem {...allProps} />);
      expect(cutTextContent).toHaveBeenCalledTimes(1);
      expect(onItemFn).not.toHaveBeenCalled();
      expect(onQuickViewFn).not.toHaveBeenCalled();

      component.find(Link).simulate('click');
      expect(onItemFn).toHaveBeenCalledTimes(1);
      expect(onQuickViewFn).not.toHaveBeenCalled();

      component.find('.quick-view-button').simulate('click');
      expect(onItemFn).toHaveBeenCalledTimes(1);
      expect(onQuickViewFn).toHaveBeenCalledTimes(1);

      cutTextContent.mockClear();
      expect(cutTextContent).toHaveBeenCalledTimes(0);

      // dispatch window event resize
      const resizeEvent = document.createEvent('Event');
      resizeEvent.initEvent('resize', true, true);
      window.dispatchEvent(resizeEvent);

      expect(debounce).toHaveBeenCalledTimes(1);
    });

    it('should correct render with all props', () => {
      const component = mount(<RecommendedItem {...allProps} />);
      expect(component.find(Link)).toHaveLength(1);
      expect(component.find(Price)).toHaveLength(2);
      expect(component.find('img')).toHaveLength(1);
      expect(component.find('.quick-view-button')).toHaveLength(1);
      expect(component.find('.quick-view-button').props().onClick).toEqual(onQuickViewFn);
      expect(component.find('.quick-view-button').props().onMouseEnter)
        .toEqual(onHoverQuickViewButtonFn);
      expect(component.find('.quick-view-button').props().onMouseLeave)
        .toEqual(onLeaveQuickViewButtonFn);
      expect(component).toMatchSnapshot();
    });

    it('should correct render without oldPrice and onQuickViewClick', () => {
      const component = mount(<RecommendedItem {...props} />);
      expect(component.find(Link)).toHaveLength(1);
      expect(component.find(Price)).toHaveLength(1);
      expect(component.find('img')).toHaveLength(1);
      expect(component.find('.quick-view-button')).toHaveLength(1);
      expect(component.find('.quick-view-button').props().onClick).toBe(undefined);
      expect(component).toMatchSnapshot();
    });

    it('onItemClick and calls correctly', () => {
      onItemFn.mockClear();
      const component = mount(<RecommendedItem {...allProps} />);
      expect(onItemFn).not.toHaveBeenCalled();

      component.find(Link).simulate('click');
      expect(onItemFn).toHaveBeenCalledTimes(1);
    });
  });

  describe('without props', () => {
    it('should check render without props', () => {
      const component = mount(<RecommendedItem />);
      expect(component.find(Link)).toHaveLength(1);
      expect(component.find(Price)).toHaveLength(0);
      expect(component.find('img')).toHaveLength(0);
      expect(component.find('.quick-view-button')).toHaveLength(0);
    });
    it('should match snapshot', () => {
      const component = shallow(<RecommendedItem />);
      expect(component).toMatchSnapshot();
    });
  });
});
