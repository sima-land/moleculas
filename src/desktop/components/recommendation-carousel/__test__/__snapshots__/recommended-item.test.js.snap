// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecommendedItem /> render with props should correct render with all props 1`] = `
<withGlobalListeners(RecommendedItem)
  addGlobalListener={[MockFunction]}
  currencyGrapheme="$"
  imageSrc="https://cdn2.static1-sima-land.com/items/3592229/4/140.jpg?v=1549604912"
  name="Милая картина из шерсти Schrodinger cat, А5"
  oldPrice={234.9}
  onItemClick={[MockFunction]}
  onQuickViewClick={[MockFunction]}
  price={135.4}
  url="https://www.sima-land.ru/3266880/nabor-dlya-tvorchestva-topiariy-xyezhnost-d-shara-6-cm/"
>
  <RecommendedItem
    addGlobalListener={[Function]}
    currencyGrapheme="$"
    imageSrc="https://cdn2.static1-sima-land.com/items/3592229/4/140.jpg?v=1549604912"
    name="Милая картина из шерсти Schrodinger cat, А5"
    oldPrice={234.9}
    onItemClick={[MockFunction]}
    onQuickViewClick={[MockFunction]}
    price={135.4}
    url="https://www.sima-land.ru/3266880/nabor-dlya-tvorchestva-topiariy-xyezhnost-d-shara-6-cm/"
  >
    <div
      className="item"
    >
      <div
        className="image-wrapper"
      >
        <img
          alt="Милая картина из шерсти Schrodinger cat, А5"
          className="image"
          src="https://cdn2.static1-sima-land.com/items/3592229/4/140.jpg?v=1549604912"
        />
        <ItemQuickViewButton
          additionalClassName="quick-view-button"
          onClick={[MockFunction]}
        >
          <button
            className="quick-view-button quick-view-button"
            onClick={[MockFunction]}
          >
            Быстрый просмотр
          </button>
        </ItemQuickViewButton>
      </div>
      <div
        className="price-wrapper"
      >
        <Price
          className="price"
          currencyGrapheme="$"
          fractionalSign=","
          value={135.4}
        >
          <span
            className="price price"
          >
            135
            ,40
            <span>
               $
            </span>
          </span>
        </Price>
        <Price
          className="old-price"
          currencyGrapheme="$"
          old={true}
          value={234.9}
        >
          <span
            className="price old-price old-price"
          >
            234
            ,90
            <span>
               $
            </span>
          </span>
        </Price>
      </div>
      <ForwardRef(Link)
        className="link"
        color="black"
        disableHoverEffect={true}
        href="https://www.sima-land.ru/3266880/nabor-dlya-tvorchestva-topiariy-xyezhnost-d-shara-6-cm/"
        onClick={[MockFunction]}
      >
        <a
          className="link link color__black"
          disableHoverEffect={true}
          href="https://www.sima-land.ru/3266880/nabor-dlya-tvorchestva-topiariy-xyezhnost-d-shara-6-cm/"
          onClick={[MockFunction]}
        >
          Милая картина из шерсти Schrodinger cat, А5
        </a>
      </ForwardRef(Link)>
    </div>
  </RecommendedItem>
</withGlobalListeners(RecommendedItem)>
`;

exports[`<RecommendedItem /> render with props should correct render without oldPrice and onQuickViewClick 1`] = `
<withGlobalListeners(RecommendedItem)
  addGlobalListener={[MockFunction]}
  currencyGrapheme="$"
  imageSrc="https://cdn2.static1-sima-land.com/items/3592229/4/140.jpg?v=1549604912"
  name="Милая картина из шерсти Schrodinger cat, А5"
  onItemClick={[MockFunction]}
  price={135.4}
  url="https://www.sima-land.ru/3266880/nabor-dlya-tvorchestva-topiariy-xyezhnost-d-shara-6-cm/"
>
  <RecommendedItem
    addGlobalListener={[Function]}
    currencyGrapheme="$"
    imageSrc="https://cdn2.static1-sima-land.com/items/3592229/4/140.jpg?v=1549604912"
    name="Милая картина из шерсти Schrodinger cat, А5"
    onItemClick={[MockFunction]}
    price={135.4}
    url="https://www.sima-land.ru/3266880/nabor-dlya-tvorchestva-topiariy-xyezhnost-d-shara-6-cm/"
  >
    <div
      className="item"
    >
      <div
        className="image-wrapper"
      >
        <img
          alt="Милая картина из шерсти Schrodinger cat, А5"
          className="image"
          src="https://cdn2.static1-sima-land.com/items/3592229/4/140.jpg?v=1549604912"
        />
        <ItemQuickViewButton
          additionalClassName="quick-view-button"
        >
          <button
            className="quick-view-button quick-view-button"
            onClick={null}
          >
            Быстрый просмотр
          </button>
        </ItemQuickViewButton>
      </div>
      <div
        className="price-wrapper"
      >
        <Price
          className="price"
          currencyGrapheme="$"
          fractionalSign=","
          value={135.4}
        >
          <span
            className="price price"
          >
            135
            ,40
            <span>
               $
            </span>
          </span>
        </Price>
      </div>
      <ForwardRef(Link)
        className="link"
        color="black"
        disableHoverEffect={true}
        href="https://www.sima-land.ru/3266880/nabor-dlya-tvorchestva-topiariy-xyezhnost-d-shara-6-cm/"
        onClick={[MockFunction]}
      >
        <a
          className="link link color__black"
          disableHoverEffect={true}
          href="https://www.sima-land.ru/3266880/nabor-dlya-tvorchestva-topiariy-xyezhnost-d-shara-6-cm/"
          onClick={[MockFunction]}
        >
          Милая картина из шерсти Schrodinger cat, А5
        </a>
      </ForwardRef(Link)>
    </div>
  </RecommendedItem>
</withGlobalListeners(RecommendedItem)>
`;

exports[`<RecommendedItem /> without props should match snapshot 1`] = `
<RecommendedItem
  addGlobalListener={[Function]}
/>
`;
